Feature/expose t3dr parameters

PerrineAguiar opened the pull request:
This PR exposes Tango 3d Reconstruction parameters as ros params.

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_node.h:
```diff
@@ -90,7 +95,12 @@ const std::string SAVE_MAP_SERVICE_NAME = "save_map";
+const bool TANGO_3DR_DEFAULT_USE_SPACE_CLEARING = false;
+const int32_t TANGO_3DR_DEFAULT_MIN_NUM_VERTICES = 1;
+const int32_t TANGO_3DR_DEFAULT_UPDATE_METHOD = 0; // TRAVERSAL_UPDATE
+const int32_t TANGO_3DR_DEFAULT_MAX_VOXEL_WEIGHT = 16383; // Roughly number of observations.
```
Where does this number come from? What kind of observations? Is it a sensor dependent value? Please add some more useful documentation to explain this.

smits replied:
This comes straight from the Tango Developers documentation: "The maximum voxel weight for the integration filter. Defaults to 16383. Roughly corresponds to the number of observations. A higher value gives better geometric quality and stability; a lower value gives more responsiveness to dynamic objects."

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_node.h:
```diff
@@ -90,7 +95,12 @@ const std::string SAVE_MAP_SERVICE_NAME = "save_map";
-const double TANGO_3D_RECONSTRUCTION_DEFAULT_RESOLUTION = 0.05; // meter
+const double TANGO_3DR_DEFAULT_RESOLUTION = 0.05; // meter
+const bool TANGO_3DR_DEFAULT_USE_SPACE_CLEARING = false;
+const int32_t TANGO_3DR_DEFAULT_MIN_NUM_VERTICES = 1;
```
Why one and not zero?

smits replied:
It's the default, we should only differ if we have a good reason.
From the Tango developers site: 

> Minimum number of vertices that every connected component in the mesh should consist of. Defaults to 1. Smaller components (typically noise) will be deleted.

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -219,9 +219,29 @@ void TangoRosNode::onInit() {
-  if (!node_handle_.hasParam(TANGO_3D_RECONSTRUCTION_RESOLUTION_PARAM_NAME)) {
-    node_handle_.setParam(TANGO_3D_RECONSTRUCTION_RESOLUTION_PARAM_NAME,
-                          TANGO_3D_RECONSTRUCTION_DEFAULT_RESOLUTION);
+  if (!node_handle_.hasParam(TANGO_3DR_RESOLUTION_PARAM_NAME)) {
```
Could you add a helper function for this to avoid typos?
`set_default_value_if_parameter_does_not_exist()`

smits replied:
I was thinking the same thing, but it will have to be templated :(

smits replied:
It will have to be templated but :+1:

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_node.h:
```diff
@@ -78,10 +78,15 @@ const std::string LOCALIZATION_MAP_UUID_PARAM_NAME = "localization_map_uuid";
+const std::string TANGO_3DR_MIN_NUM_VERTICES_PARAM_NAME = "reconstruction/min_num_vertices";
+const std::string TANGO_3DR_UPDATE_METHOD_PARAM_NAME = "reconstruction/update_method";
+const std::string TANGO_3DR_MAX_VOXEL_WEIGHT_PARAM_NAME = "reconstruction/max_voxel_weight";
+const std::string TANGO_3DR_FLOORPLAN_MAX_ERROR_PARAM_NAME = "reconstruction/floorplan_max_error";
```
This class is again getting really big. Can you move all 3D reconstruction related code be moved to a separate helper class?

tango_ros_3d_reconstruction.h
```
(...) // parameter names
(...) // default values
void setupTango3drConfig(...)
```

Extra advantage: This allows to easily disable the 3DR reconstruction code based on the value of `enable_depth_camera` to support use cases without depth camera.

smits replied:
I guess collecting them in struct could do as well?

smits commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -476,6 +496,62 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+    LOG(ERROR) << function_name << ", Tango3DR_Config_setBool "
+        << use_space_clearing << " error: " << result;
+    return result;
+  }
```
It would be really nice if we could wrap this as well, maybe if we store the parameters per type in a map
```
struct bool_param {
  const char* name;
  std::string param_name;
  bool default value;
  setRosParamToTangoConfig(ros::NodeHandle& nh, Tango3DRConfig& t3dr_config) {
   bool value;
    node_handle_.param(param_name,
                      value, default_value);
   result = Tango3DR_Config_setBool(t3dr_config, name,
                                    value);
   if (result != TANGO_3DR_SUCCESS) {
     LOG(ERROR) << "setRosParamToTangoConfig, Tango3DR_Config_setBool "
         << name << " error: " << result;
     return result;
   }
}
```
The same for int32 and double parameters.
put these in a set:
`std::forward_list<bool_param> bool_params;`
```
for ( const bool_param& param : bool_params) {
  param.setRosParamToTangoConfig(node_handle_, t3dr_config)
}
```

I'm not sure if it is worth the effort though.

PerrineAguiar replied:
Should I do this in this PR or open an issue about it?

PerrineAguiar commented:
@mcopejans @smits Thanks for the review. I addressed your comments.

smits commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_3d_reconstruction_helper.h:
```diff
@@ -0,0 +1,63 @@
+         Tango3DR_Pose* last_camera_color_pose,
+         Tango3DR_GridIndexArray* t3dr_updated_indices);
+
+void ExtractMeshAndConvertItTMarkerArray(
```
rename to `ExtractMeshAndConvertToMarkerArray`?

smits commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_3d_reconstruction_helper.h:
```diff
@@ -0,0 +1,63 @@
+    const Tango3DR_GridIndexArray& t3dr_updated_indices,
+    double time_offset, visualization_msgs::MarkerArray* mesh_marker_array);
+
+bool ExtractFloorPlanImageAndConvertItToOccupancyGrid(
```
`ExtractFloorPlanImageAndConvertToOccupancyGrid`?

PerrineAguiar commented:
@smits Thank you, I addressed your comments.
I changed some function names from camelCase to CamelCase for consistency with the rest of the functions.

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_3d_reconstruction_helper.h:
```diff
@@ -0,0 +1,63 @@
+namespace tango_3d_reconstruction_helper {
+const double TANGO_3DR_DEFAULT_RESOLUTION = 0.05; // meter
+const bool TANGO_3DR_DEFAULT_USE_SPACE_CLEARING = false;
+const int32_t TANGO_3DR_DEFAULT_MIN_NUM_VERTICES = 1; // Default value from TangoCongig
```
Typo `Congig -> Config`

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_3d_reconstruction_helper.h:
```diff
@@ -0,0 +1,63 @@
+const bool TANGO_3DR_DEFAULT_USE_SPACE_CLEARING = false;
+const int32_t TANGO_3DR_DEFAULT_MIN_NUM_VERTICES = 1; // Default value from TangoCongig
+const int32_t TANGO_3DR_DEFAULT_UPDATE_METHOD = 0; // TRAVERSAL_UPDATE
+const int32_t TANGO_3DR_DEFAULT_MAX_VOXEL_WEIGHT = 16383; // Default value from TangoCongig
```
Same

PerrineAguiar commented:
@smits @mcopejans Thanks for your review.
@mcopejans When the depth camera is disabled the topic publishing the occupancy grid is not advertised. Therefore if you subscribe to it, you won't get any data.
